### PR TITLE
Add support for repo-level commit exclusion #246

### LINKED ISSUE #246: Add back-end support for repo-level commit exclusion
Assuming config component is able to specify a list of commit hashes to be excluded from analysis, prepare the backend to honor that request.

### PR DESCRIPTION
**Pull request by yamidark**
Fixes #246
```
User may sometimes wish to ignore some 'ugly' commits
made that will distort the dashboard's rampchart statistics.

To allow this, let's add support for repo-level commit
exclusions.
```

### CODE REVIEW

**Comment by yamidark** [comment]
@eugenepeh @ongspxm @yong24s For your reviews 1st, will add the tests later on

File: src/main/java/reposense/system/CommandRunner.java [hunk 1]
```diff
@@ -108,6 +112,24 @@ public static String getCommitHashBeforeDate(String root, String branchName, Dat
...
+    public static void removeCommitAuthor(String root, String branch, String commitHash) {
+        Path rootPath = Paths.get(root);
+        // create temp branch with the given commitHash as the HEAD
+        runCommand(rootPath, String.format(CHECKOUT_COMMIT_IN_TEMP_BRANCH_COMMAND_FORMAT, commitHash));
+
+        // remove the author for the given commit
+        runCommand(rootPath, REMOVE_COMMIT_AUTHOR_COMMAND);
+
+        // rebase temp branch onto the given branch, replacing the given commit's author on that branch
+        runCommand(rootPath, String.format(REBASE_ONTO_MASTER_FROM_TEMP_BRANCH_COMMAND_FORMAT, branch));
```
Comment (eugenepeh) [inline src/main/java/reposense/system/CommandRunner.java]: shouldn't it be rebase master onto the given branch?

File: src/main/java/reposense/authorship/AuthorshipReporter.java [hunk 2]
```diff
@@ -18,6 +19,8 @@
      * Generates and returns the authorship summary for each repo in {@code config}.
      */
     public static AuthorshipSummary generateAuthorshipSummary(RepoConfiguration config) {
+        removeAuthorForIgnoredCommits(config);
```
Comment (eugenepeh) [inline src/main/java/reposense/authorship/AuthorshipReporter.java]: removeAuthor~For~`From`IgnoredCommits

File: src/main/java/reposense/system/CommandRunner.java [hunk 3]
```diff
@@ -108,6 +112,24 @@ public static String getCommitHashBeforeDate(String root, String branchName, Dat
         return runCommand(rootPath, revListCommand);
     }
 
+    /**
+     * Removes the author for the {@code commitHash} in the {@code branch} of the repo.
+     */
+    public static void removeCommitAuthor(String root, String branch, String commitHash) {
+        Path rootPath = Paths.get(root);
+        // create temp branch with the given commitHash as the HEAD
+        runCommand(rootPath, String.format(CHECKOUT_COMMIT_IN_TEMP_BRANCH_COMMAND_FORMAT, commitHash));
```
Comment (eugenepeh) [inline src/main/java/reposense/system/CommandRunner.java]: note: this command will fail if a `temp` branch already exists. we should attempt to delete `temp` branch first or use a unique branch name

File: src/main/java/reposense/system/CommandRunner.java (hunk 1 above)
Comment (yamidark) [reply src/main/java/reposense/system/CommandRunner.java]: nope, we are doing the editing on the given branch 1st, then rebasing those changes into master branch, as the master branch that we edit is temporary, and won't affect the original.

File: src/main/java/reposense/system/CommandRunner.java (hunk 3 above)
Comment (yamidark) [reply src/main/java/reposense/system/CommandRunner.java]: Hmmm, your right, perhaps we can use the `commitHash` as the branch name instead, highly doubt any1 would put their branch names as that xD

File: src/main/java/reposense/report/ReportGenerator.java [hunk 4]
```diff
@@ -40,7 +41,7 @@
      * @throws IOException if templateZip.zip does not exists in jar file.
      */
     public static void generateReposReport(List<RepoConfiguration> configs, String outputPath,
-            String generationDate) throws IOException {
+                                           String generationDate) throws IOException {
```
Comment (eugenepeh) [inline src/main/java/reposense/report/ReportGenerator.java]: indentation

File: src/main/java/reposense/system/CommandRunner.java [hunk 5]
```diff
@@ -21,6 +21,11 @@
     // ignore check against email
     private static final String AUTHOR_NAME_PATTERN = "^%s <.*>$";
     private static final String OR_OPERATOR_PATTERN = "\\|";
+    private static final String REBASE_ONTO_BRANCH_FROM_TEMP_BRANCH_COMMAND_FORMAT = "git rebase --onto %s HEAD %s";
+    private static final String REMOVE_COMMIT_AUTHOR_COMMAND =
+            "git commit --amend --author=\"- <temp@github.com>\" --no-edit";
```
Comment (eugenepeh) [inline src/main/java/reposense/system/CommandRunner.java]: the email can be changed back to `-`?

File: src/main/java/reposense/report/ReportGenerator.java (hunk 4 above)
Comment (yong24s) [inline src/main/java/reposense/report/ReportGenerator.java]: 8 spaces indentation

File: src/main/java/reposense/report/ReportGenerator.java [hunk 6]
```diff
@@ -103,6 +105,15 @@ private static void generateIndividualRepoReport(
         FileUtil.writeJsonFile(authorshipSummary.getFileResults(), getIndividualAuthorshipPath(repoReportDirectory));
     }
 
+    /**
+     * Removes all the author names for the commits that we are ignoring in {@code config}
```
Comment (yong24s) [inline src/main/java/reposense/report/ReportGenerator.java]: .

File: src/test/java/reposense/system/CommandRunnerTest.java [hunk 7]
```diff
@@ -180,6 +181,46 @@ public void getCommitHashBeforeDate_invalidBranch_throwsRunTimeException() {
         CommandRunner.getCommitHashBeforeDate(config.getRepoRoot(), "invalidBranch", date);
     }
 
+    @Test
+    public void removeCommitAuthor_singleValidCommitHash_success() {
+        // create new temp branch so as to not affect the master branch for other tests
+        CommandRunner.checkoutNewBranch(config.getRepoRoot(), TEMP_BRANCH);
```
Comment (eugenepeh) [inline src/test/java/reposense/system/CommandRunnerTest.java]: why don't just use the `removeAuthorFromIgnoredCommits` method?

Comment (yamidark) [reply src/test/java/reposense/system/CommandRunnerTest.java]: its a unit test within `CommandRunner`

**Comment by eugenepeh** [comment]
Removing this from v1.2 as it doesn't make much changes.
Schedule to next milestone

File: src/test/java/reposense/system/CommandRunnerTest.java (hunk 7 above)
Comment (eugenepeh) [reply src/test/java/reposense/system/CommandRunnerTest.java]: ic. unit test is good but testing the `removeAuthorFromIgnoredCommits` is more important.
do include it in the next iteration.

**Comment by yamidark** [comment]
Our repository over removing some of the 'large' commits:

screenshot from 2018-08-14 20-34-44

File: docs/UserGuide.md [hunk 8]
```diff
@@ -99,8 +99,9 @@ Column Name | Explanation
 ----------- | -----------
 Repository's Location | The `GitHub URL` or `Disk Path` to the git repository
 Branch | The branch to analyse in the target repository
-[Optional] Ignore Global List | The list of file path globs to ignore during analysis for each author. More details on the Java glob standard [here](https://javapapers.com/java/glob-with-java-nio/)
+[Optional] Ignore Glob List | The list of file path globs to ignore during analysis for each author. More details on the Java glob standard [here](https://javapapers.com/java/glob-with-java-nio/)
 [Optional] Ignore standalone config | Ignore the presence of standalone config in target repository. To do so, enter **`yes`** in this column. Otherwise, the configuration in the target repository will be used by default.
+[Optional] Ignore Commit List | The list of commits to ignore during analysis. For accurate results, provide the commits with their full hash in reverse chronological order.
```
Comment (eugenepeh) [inline docs/UserGuide.md]: description doesn't match the StandaloneConfiguration.md example.

File: docs/UserGuide.md [hunk 9]
```diff
@@ -175,8 +176,7 @@ You can use `Start Tags` to mark the start of your contribution. The author spec
 
 #### Standalone Configuration
 
-Instead of self-configuring all the repository details, you can have the repository owners include a 
-`Standalone Configuration` to maintain their own repository configuration. For more information, check out the
+Instead of self-configuring all the repository details, you can have the repository owners include a `Standalone Configuration` to maintain their own repository configuration. For more information, check out the
```
Comment (eugenepeh) [inline docs/UserGuide.md]: why this change?

File: src/functional/resources/repo-config.csv [hunk 10]
```diff
@@ -1,4 +1,4 @@
-Repository's Location,Branch,Ignore Glob List,Ignore standalone config
-https://github.com/reposense/testrepo-Beta.git,master,docs**,yes
-https://github.com/reposense/testrepo-Charlie.git,master,,
-https://github.com/reposense/testrepo-Delta.git,master,,
+Repository's Location,Branch,Ignore Glob List,Ignore standalone config,Ignore Commits List
+https://github.com/reposense/testrepo-Beta.git,master,docs**,yes,
+https://github.com/reposense/testrepo-Charlie.git,master,,,
+https://github.com/reposense/testrepo-Delta.git,master,,,
```
Comment (eugenepeh) [inline src/functional/resources/repo-config.csv]: we need to include it in functional test as well. Let's use charlie (in this config) and delta (standalone).

File: src/main/java/reposense/git/GitCommitEditer.java [hunk 11]
```diff
@@ -0,0 +1,19 @@
+package reposense.git;
+
+import reposense.model.RepoConfiguration;
+import reposense.system.CommandRunner;
+
+/**
+ * Executes git commands that will edit the repository's commit information.
```
Comment (eugenepeh) [inline src/main/java/reposense/git/GitCommitEditer.java]: ~that will~ -> `to` since you're doing it.

File: src/main/java/reposense/git/GitCommitEditer.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+ */
+public class GitCommitEditer {
+
+    /**
+     * Removes all the author names for the commits that we are ignoring in {@code config}.
+     * Assumes that the list of commit hashes in {@code config} is given in reverse-chronological order.
```
Comment (eugenepeh) [inline src/main/java/reposense/git/GitCommitEditer.java]: Why the assumption when this isn't compulsory as written in docs?

File: src/main/java/reposense/git/GitCommitEditer.java [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+import reposense.model.RepoConfiguration;
+import reposense.system.CommandRunner;
+
+/**
+ * Executes git commands that will edit the repository's commit information.
+ */
+public class GitCommitEditer {
+
+    /**
+     * Removes all the author names for the commits that we are ignoring in {@code config}.
```
Comment (eugenepeh) [inline src/main/java/reposense/git/GitCommitEditer.java]: ~that we are ignoring~ -> `were specified to ignore`

File: src/test/java/reposense/git/GitCommitEditerTest.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+public class GitCommitEditerTest extends GitTestTemplate {
+    private static final int NUMBER_FAKE_AUTHOR_COMMIT = 4;
+    private static final int NUMBER_EUGENE_AUTHOR_COMMIT = 1;
+
+    @Before
+    @Override
+    public void before() throws InvalidLocationException {
+        super.before();
+        // create new temp branch so as to not affect the master branch for other tests
+        CommandRunner.checkoutNewBranch(config.getRepoRoot(), TEMP_BRANCH);
```
Comment (eugenepeh) [inline src/test/java/reposense/git/GitCommitEditerTest.java]: perhaps we should do this in the `GitTestTemplate` itself?

File: src/test/java/reposense/template/GitTestTemplate.java [hunk 15]
```diff
@@ -32,6 +32,11 @@
     protected static final String MAIN_AUTHOR_NAME = "harryggg";
     protected static final String FAKE_AUTHOR_NAME = "fakeAuthor";
     protected static final String EUGENE_AUTHOR_NAME = "eugenepeh";
+    protected static final String LATEST_EUGENE_AUTHOR_COMMIT_HASH = "2d87a431fcbb8f73a731b6df0fcbee962c85c250";
```
Comment (eugenepeh) [inline src/test/java/reposense/template/GitTestTemplate.java]: latest is subjective though, it might change as we update the test repo.
How about `COMMIT_HASH_EUGENE_DD_MM_YYYY`?

File: docs/UserGuide.md (hunk 9 above)
Comment (yamidark) [reply docs/UserGuide.md]: Oh, was just cleaning up the documents, since there was trailing whitespace and a newline there, which wasn't necessary (my travis was failing earlier because of the `run_check.sh` I think, so I thought to clear those warnings as well since I'm editing the docs anyway)

File: src/functional/resources/repo-config.csv (hunk 10 above)
Comment (yamidark) [reply src/functional/resources/repo-config.csv]: will add some 'large' commits in the functional tests then

File: src/main/java/reposense/git/GitCommitEditer.java (hunk 12 above)
Comment (yamidark) [reply src/main/java/reposense/git/GitCommitEditer.java]: Hmm, so we should rephrase it to be the same as in the docs: `For accurate results, the commits should be provided with their full hash in reverse chronological order.`?

File: src/test/java/reposense/git/GitCommitEditerTest.java (hunk 14 above)
Comment (yamidark) [reply src/test/java/reposense/git/GitCommitEditerTest.java]: Sounds good, ensure all tests don't affect the original "master" in anyway

File: src/main/java/reposense/git/GitCommitEditer.java [hunk 16]
```diff
@@ -0,0 +1,19 @@
+package reposense.git;
+
+import reposense.model.RepoConfiguration;
+import reposense.system.CommandRunner;
+
+/**
+ * Executes git commands to edit the repository's commit information.
+ */
+public class GitCommitEditer {
```
Comment (yong24s) [inline src/main/java/reposense/git/GitCommitEditer.java]: Editor?

**Comment by yamidark** [comment]
@eugenepeh @damithc  did some testing, for some reason rebasing may have conflicts due to previous merge conflicts commits even though we are just changing the author name, so I tried to 'bypass' this by setting 'Xours' or 'Xtheirs', but some of the other commits will be affected, resulting in a different final repository at the end.

Tried other solutions such as [`git filter-branch`](https://git-scm.com/docs/git-filter-branch), but using commit hashes it only allows me to edit 1 commit, or using `GIT_COMITTER_DATE` will cause all the commits to be rewritten (no idea why).

Should we continue adding this feature in, where we recommend this feature is best used without any merge conflict (using squash merge or rebase merge?) in the repository, and perhaps skip 'removing' of these commits if there are merge conflicts.

Or we could downgrade this to only remove `CommitInfo`, not affecting `AuthorInfo`?

**Comment by eugenepeh** [comment]
> Perhaps we can put it on hold until we figure out a way around these problems?

Another way would be to perform the alteration in [FileInfoExtractor](https://github.com/reposense/RepoSense/blob/master/src/main/java/reposense/authorship/FileInfoExtractor.java) where if you detect the hash to be similar, modify the author information.

**Comment by yamidark** [comment]
>Another way would be to perform the alteration in FileInfoExtractor where if you detect the hash to be similar, modify the author information.

The `FileInfo` doesn't contain the hash.
Rather, I believe we can get the commit hash for the line in `git blame`, which since we currently don't have copy/move detection, should be the last commit which edited that line (excluding whitespaces). Then if that commit hash is inside the `ignoreCommitList`, we remove the author contribution.

In fact, now that I think about this carefully, I believe this will give the same result as what we are trying to achieve through 'removing authors' in the commits, since if a certain commit would have attributed a line to a author, than 'removing the commit's author' or 'remove the author if the commit is the same' does the same thing.

@damithc Your opinion on this?

**Comment by damithc** [comment]
> @damithc Your opinion on this?

I'm not familiar enough with the code to give an opinion. But I'm OK to give it a try if it gives the behavior we need.

**Comment by yamidark** [comment]
Tested with RepoSense using the old and new method, both gave the same result using the same commit hashes.

File: src/test/java/reposense/commits/CommitInfoAnalyzerTest.java [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+
+        List<CommitInfo> commitInfos = CommitInfoExtractor.extractCommitInfos(config);
+        List<CommitResult> commitResults = CommitInfoAnalyzer.analyzeCommits(commitInfos, config);
+
+        Assert.assertEquals(commitInfos.size() - NUMBER_EUGENE_COMMIT, commitResults.size());
+    }
+
+    @Test
+    public void analyzeCommits_eugeneAuthorNoIgnoredCommitsNoDateRange_success() {
+        config.getAuthorAliasMap().clear();
```
Comment (eugenepeh) [inline src/test/java/reposense/commits/CommitInfoAnalyzerTest.java]: is there a need to do this?

File: src/test/java/reposense/commits/CommitInfoAnalyzerTest.java [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    @Test
+    public void analyzeCommits_allAuthorSingleCommitIgnoredNoDateRange_success() {
+        config.getAuthorAliasMap().put(MAIN_AUTHOR_NAME, new Author(MAIN_AUTHOR_NAME));
+        config.getAuthorAliasMap().put(FAKE_AUTHOR_NAME, new Author(FAKE_AUTHOR_NAME));
+        config.getAuthorAliasMap().put(EUGENE_AUTHOR_NAME, new Author(EUGENE_AUTHOR_NAME));
+        List<CommitInfo> commitInfos = CommitInfoExtractor.extractCommitInfos(config);
+
+        config.setIgnoreCommitList(Collections.singletonList(FAKE_AUTHOR_BLAME_TEST_FILE_COMMIT_08022018));
```
Comment (eugenepeh) [inline src/test/java/reposense/commits/CommitInfoAnalyzerTest.java]: since this belongs to setting up the config, should be place tgt on top.

File: src/test/java/reposense/commits/CommitInfoAnalyzerTest.java [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    @Test
+    public void analyzeCommits_allAuthorMultipleCommitIgnoredNoDateRange_success() {
+        config.getAuthorAliasMap().put(MAIN_AUTHOR_NAME, new Author(MAIN_AUTHOR_NAME));
+        config.getAuthorAliasMap().put(FAKE_AUTHOR_NAME, new Author(FAKE_AUTHOR_NAME));
+        config.getAuthorAliasMap().put(EUGENE_AUTHOR_NAME, new Author(EUGENE_AUTHOR_NAME));
+        List<CommitInfo> commitInfos = CommitInfoExtractor.extractCommitInfos(config);
+
+        config.setIgnoreCommitList(
```
Comment (eugenepeh) [inline src/test/java/reposense/commits/CommitInfoAnalyzerTest.java]: same for this.

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 20]
```diff
@@ -31,6 +31,10 @@
     private static final Pattern GIT_REPOSITORY_LOCATION_PATTERN =
             Pattern.compile("^.*github.com\\/(?<org>.+?)\\/(?<repoName>.+?)\\.git$");
     private static final String DEFAULT_BRANCH = "master";
+    private static final String COMMIT_HASH_REGEX = "^[0-9a-f]*$";
```
Comment (eugenepeh) [inline src/main/java/reposense/model/RepoConfiguration.java]: hmm, your commits matching uses startsWith. So if someone puts empty string into ignore commits, it will basically ignores everything?

maybe should change this to `+` instead.

File: src/functional/resources/repo-config.csv [hunk 21]
```diff
@@ -1,6 +1,6 @@
-Repository's Location,Branch,Ignore Glob List,Ignore standalone config
-https://github.com/reposense/testrepo-Alpha.git,master,,
-https://github.com/reposense/testrepo-Beta.git,master,docs**,yes
-https://github.com/reposense/testrepo-Charlie.git,master,,
-https://github.com/reposense/testrepo-Delta.git,master,,
-https://github.com/reposense/testrepo-Delta.git,nonExistentBranch,,
+Repository's Location,Branch,Ignore Glob List,Ignore standalone config,Ignore Commits List
+https://github.com/reposense/testrepo-Alpha.git,master,,,
```
Comment (eugenepeh) [inline src/functional/resources/repo-config.csv]: a bit hard to visually verify the results... can we add hashes to make fakeAuthor empty in both functional n netlify?

File: docs/UserGuide.md (hunk 9 above)
Comment (damithc) [reply docs/UserGuide.md]: Use _Standalone Configuration_ instead of `Standalone Configuration` (because this is simply a new term, not a reference to something in the code)

File: docs/StandaloneConfiguration.md [hunk 22]
```diff
@@ -25,13 +26,16 @@
...
+"ignoreCommitList": ["12345abc", "67890def"] <-- The list of commits to ignore during analysis. For accurate
+                                           results, the commits should be provided with their full hash.
+
+"ignoreGlobList": ["**.css"]                 <-- Author level's ignoreGlobList adds on to the Repository level's.
                                            Thus, the actual ignoreGlobList for alice would contain
                                            css, dat and js.
 
-"githubId": "bob"                      <-- Only githubId is mandatory.
+"githubId": "bob"                            <-- Only githubId is mandatory.
                                            Optional information can be left out to prevent clutter.
```
Comment (yong24s) [inline docs/StandaloneConfiguration.md]: Minor nits. No need for change.
Other than arrows, text should be aligned with respective to its block.

```
<-- Only githubId is mandatory.
    Optional information can be left out to prevent clutter.
```
